File: agents/macos/Sources/N3NAgentCore/Protocol/SecureMessageService.swift
import Foundation
import CryptoKit
import Logging

/// Service for encrypting and decrypting agent messages.
public final class SecureMessageService {

    private let crypto = AgentCrypto.shared
    private let storage = KeychainStorage.shared
    private let logger = Logger(label: "n3n.agent.secure-message")

    /// Maximum time drift allowed (5 minutes in milliseconds)
    private let maxTimeDriftMs: Int64 = 5 * 60 * 1000

    /// Current sequence counter for outgoing messages
    private var outgoingSequence: Int64

    public init() {
        self.outgoingSequence = Int64(Date().timeIntervalSince1970 * 1000)
    }

    // MARK: - Encryption

    /// Encrypt a message to send to the server.
    public func encrypt<T: Encodable>(_ payload: T, deviceKeys: DeviceKeys) throws -> String {
        // Serialize payload
        let encoder = JSONEncoder()
        let plaintext = try encoder.encode(payload)

        // Get encryption key (client to server)
        let encryptKey = SymmetricKey(data: deviceKeys.encryptKeyC2S)

        // Build initial header (nonce will be set after encryption)
        let nonce = AES.GCM.Nonce()
        let header = SecureMessage.Header(
            did: deviceKeys.deviceId,
            seq: nextSequence(),
            nonce: Data(nonce).base64URLEncodedString(),
            dir: "c2s"
        )

        // Get header bytes for AAD
        let headerBytes = try encoder.encode(header)

        // Encrypt with AAD
        let sealedBox = try AES.GCM.seal(plaintext, using: encryptKey, nonce: nonce, authenticating: headerBytes)

        // Build secure message
        let message = SecureMessage(
            header: header,
            ciphertext: sealedBox.ciphertext.base64URLEncodedString(),
            tag: sealedBox.tag.base64URLEncodedString()
        )

        return try message.toCompact()
    }

    // MARK: - Decryption

    /// Decrypt and verify a message from the server.
    public func decrypt<T: Decodable>(
        _ encryptedMessage: String,
        as type: T.Type,
        deviceKeys: DeviceKeys
    ) throws -> DecryptedMessage<T> {
        // Parse message
        let message = try SecureMessage.fromCompact(encryptedMessage)
        let header = message.header

        // 1. Validate protocol version
        guard header.v == 1 else {
            throw SecureMessageError.invalidFormat
        }

        // 2. Validate device ID
        guard header.did == deviceKeys.deviceId else {
            throw SecureMessageError.unknownDevice
        }

        // 3. Validate timestamp
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let timeDrift = abs(now - header.ts)
        guard timeDrift <= maxTimeDriftMs else {
            logger.warning("Message expired: drift=\(timeDrift)ms")
            throw SecureMessageError.messageExpired
        }

        // 4. Validate sequence (for replay protection)
        guard header.seq > deviceKeys.lastSequence else {
            logger.warning("Replay detected: seq=\(header.seq), last=\(deviceKeys.lastSequence)")
            throw SecureMessageError.replayDetected
        }

        // 5. Get decryption key (server to client)
        let decryptKey = SymmetricKey(data: deviceKeys.encryptKeyS2C)

        // 6. Get AAD (header bytes)
        let encoder = JSONEncoder()
        let headerBytes = try encoder.encode(header)

        // 7. Decrypt
        let nonce = try AES.GCM.Nonce(data: message.getNonceBytes())
        let sealedBox = try AES.GCM.SealedBox(
            nonce: nonce,
            ciphertext: message.getCiphertextBytes(),
            tag: message.getTagBytes()
        )

        let plaintext = try AES.GCM.open(sealedBox, using: decryptKey, authenticating: headerBytes)

        // 8. Parse payload
        let decoder = JSONDecoder()
        let payload = try decoder.decode(T.self, from: plaintext)

        logger.debug("Successfully decrypted message from server")

        return DecryptedMessage(
            deviceId: header.did,
            timestamp: header.ts,
            sequence: header.seq,
            payload: payload
        )
    }

    /// Verify message without decrypting (for initial validation)
    public func verify(_ encryptedMessage: String, deviceKeys: DeviceKeys) -> VerificationResult {
        do {
            let message = try SecureMessage.fromCompact(encryptedMessage)
            let header = message.header

            // Basic validation
            guard header.v == 1 else {
                return .invalid("Unsupported protocol version")
            }

            // Check device ID
            guard header.did == deviceKeys.deviceId else {
                return .invalid("Device ID mismatch")
            }

            // Check timestamp
            let now = Int64(Date().timeIntervalSince1970 * 1000)
            let timeDrift = abs(now - header.ts)
            guard timeDrift <= maxTimeDriftMs else {
                return .invalid("Message expired")
            }

            // Check sequence
            guard header.seq > deviceKeys.lastSequence else {
                return .invalid("Replay detected")
            }

            return .valid(deviceId: header.did, sequence: header.seq)

        } catch {
            return .invalid("Parse error: \(error.localizedDescription)")
        }
    }

    // MARK: - Helpers

    private func nextSequence() -> Int64 {
        outgoingSequence += 1
        return outgoingSequence
    }
}

// MARK: - Result Types

public struct DecryptedMessage<T> {
    public let deviceId: String
    public let timestamp: Int64
    public let sequence: Int64
    public let payload: T
}

public enum VerificationResult {
    case valid(deviceId: String, sequence: Int64)
    case invalid(String)

    public var isValid: Bool {
        if case .valid = self { return true }
        return false
    }
}

File: src/main/java/com/aiinpocket/n3n/gateway/security/SecureMessageService.java
package com.aiinpocket.n3n.gateway.security;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.time.Instant;
import java.util.Base64;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Service for encrypting and decrypting agent messages.
 */
@Service
@Slf4j
@RequiredArgsConstructor
public class SecureMessageService {

    private static final long MAX_TIME_DRIFT_MS = 5 * 60 * 1000; // 5 minutes

    private final AgentCrypto agentCrypto;
    private final DeviceKeyStore deviceKeyStore;
    private final ObjectMapper objectMapper;

    // Sequence counter for outgoing messages
    private final AtomicLong outgoingSequence = new AtomicLong(System.currentTimeMillis());

    /**
     * Encrypt a message to send to a device.
     */
    public String encrypt(String deviceId, Object payload) throws SecureMessageException {
        DeviceKeyStore.DeviceKey deviceKey = deviceKeyStore.getDeviceKey(deviceId)
            .orElseThrow(() -> new SecureMessageException("Unknown device: " + deviceId));

        if (deviceKey.isRevoked()) {
            throw new SecureMessageException("Device has been revoked: " + deviceId);
        }

        try {
            // Serialize payload
            byte[] plaintext = objectMapper.writeValueAsBytes(payload);

            // Get encryption key (server to client)
            byte[] encryptKey = Base64.getDecoder().decode(deviceKey.getEncryptKeyS2C());

            // Encrypt
            AgentCrypto.EncryptedMessage encrypted = agentCrypto.encrypt(plaintext, encryptKey, null);

            // Build header
            SecureMessage.Header header = SecureMessage.Header.builder()
                .v(1)
                .alg("A256GCM")
                .did(deviceId)
                .ts(System.currentTimeMillis())
                .seq(outgoingSequence.incrementAndGet())
                .nonce(Base64.getUrlEncoder().withoutPadding().encodeToString(encrypted.nonce()))
                .dir("s2c")
                .build();

            // Re-encrypt with AAD (header)
            byte[] headerBytes = objectMapper.writeValueAsBytes(header);
            encrypted = agentCrypto.encrypt(plaintext, encryptKey, headerBytes);

            // Build secure message
            SecureMessage message = SecureMessage.builder()
                .header(header)
                .ciphertext(Base64.getUrlEncoder().withoutPadding().encodeToString(encrypted.ciphertext()))
                .tag(Base64.getUrlEncoder().withoutPadding().encodeToString(encrypted.tag()))
                .build();

            // Update header with actual nonce used
            message.getHeader().setNonce(
                Base64.getUrlEncoder().withoutPadding().encodeToString(encrypted.nonce())
            );

            return message.toCompact(objectMapper);

        } catch (Exception e) {
            throw new SecureMessageException("Encryption failed: " + e.getMessage(), e);
        }
    }

    /**
     * Decrypt and verify a message from a device.
     */
    public <T> DecryptedMessage<T> decrypt(String encryptedMessage, Class<T> payloadType)
            throws SecureMessageException {

        SecureMessage message;
        try {
            message = SecureMessage.fromCompact(encryptedMessage, objectMapper);
        } catch (Exception e) {
            throw new SecureMessageException("Invalid message format: " + e.getMessage(), e);
        }

        SecureMessage.Header header = message.getHeader();

        // 1. Validate protocol version
        if (header.getV() != 1) {
            throw new SecureMessageException("Unsupported protocol version: " + header.getV());
        }

        // 2. Get device key
        DeviceKeyStore.DeviceKey deviceKey = deviceKeyStore.getDeviceKey(header.getDid())
            .orElseThrow(() -> new SecureMessageException("Unknown device: " + header.getDid()));

        if (deviceKey.isRevoked()) {
            throw new SecureMessageException("Device has been revoked: " + header.getDid());
        }

        // 3. Validate timestamp (prevent replay with old messages)
        long now = System.currentTimeMillis();
        long timeDrift = Math.abs(now - header.getTs());
        if (timeDrift > MAX_TIME_DRIFT_MS) {
            throw new SecureMessageException("Message expired or clock drift too large: " + timeDrift + "ms");
        }

        // 4. Validate sequence (prevent replay within time window)
        if (header.getSeq() <= deviceKey.getLastSequence()) {
            throw new SecureMessageException("Replay detected: sequence " + header.getSeq() +
                " <= last " + deviceKey.getLastSequence());
        }

        try {
            // 5. Get decryption key (client to server)
            byte[] decryptKey = Base64.getDecoder().decode(deviceKey.getEncryptKeyC2S());

            // 6. Get AAD (header bytes)
            byte[] headerBytes = message.getHeaderBytes(objectMapper);

            // 7. Decrypt
            byte[] plaintext = agentCrypto.decrypt(
                message.getCiphertextBytes(),
                message.getTagBytes(),
                decryptKey,
                message.getNonceBytes(),
                headerBytes
            );

            // 8. Update sequence number
            deviceKeyStore.updateSequence(header.getDid(), header.getSeq());

            // 9. Parse payload
            T payload = objectMapper.readValue(plaintext, payloadType);

            log.debug("Successfully decrypted message from device: {}", header.getDid());

            return new DecryptedMessage<>(
                header.getDid(),
                deviceKey.getUserId(),
                header.getTs(),
                header.getSeq(),
                payload
            );

        } catch (GeneralSecurityException e) {
            throw new SecureMessageException("Decryption failed (invalid key or tampered): " + e.getMessage(), e);
        } catch (Exception e) {
            throw new SecureMessageException("Decryption failed: " + e.getMessage(), e);
        }
    }

    /**
     * Verify message authenticity without decrypting (for initial validation).
     */
    public VerificationResult verify(String encryptedMessage) {
        try {
            SecureMessage message = SecureMessage.fromCompact(encryptedMessage, objectMapper);
            SecureMessage.Header header = message.getHeader();

            // Basic validation
            if (header.getV() != 1) {
                return VerificationResult.invalid("Unsupported protocol version");
            }

            // Check device exists
            var deviceKey = deviceKeyStore.getDeviceKey(header.getDid());
            if (deviceKey.isEmpty()) {
                return VerificationResult.invalid("Unknown device");
            }

            if (deviceKey.get().isRevoked()) {
                return VerificationResult.invalid("Device revoked");
            }

            // Check timestamp
            long now = System.currentTimeMillis();
            long timeDrift = Math.abs(now - header.getTs());
            if (timeDrift > MAX_TIME_DRIFT_MS) {
                return VerificationResult.invalid("Message expired");
            }

            // Check sequence
            if (header.getSeq() <= deviceKey.get().getLastSequence()) {
                return VerificationResult.invalid("Replay detected");
            }

            return VerificationResult.valid(header.getDid(), deviceKey.get().getUserId());

        } catch (Exception e) {
            return VerificationResult.invalid("Parse error: " + e.getMessage());
        }
    }

    // Result classes

    public record DecryptedMessage<T>(
        String deviceId,
        java.util.UUID userId,
        long timestamp,
        long sequence,
        T payload
    ) {}

    public record VerificationResult(
        boolean valid,
        String deviceId,
        java.util.UUID userId,
        String error
    ) {
        public static VerificationResult valid(String deviceId, java.util.UUID userId) {
            return new VerificationResult(true, deviceId, userId, null);
        }

        public static VerificationResult invalid(String error) {
            return new VerificationResult(false, null, null, error);
        }
    }

    public static class SecureMessageException extends Exception {
        public SecureMessageException(String message) {
            super(message);
        }

        public SecureMessageException(String message, Throwable cause) {
            super(message, cause);
        }
    }
}
